### PR TITLE
GRW-1998 / Implement comparison table

### PR DESCRIPTION
**Pull request by amandaflod**
## Describe your changes

- Added the comparison table components to Offer Presenter with toggle button. 
- Updated PriceIntent query to include perils

TODO: 
- Get the short titles for the header of the table
- Order the insurances from the one with most coverage to least. 

Screenshot 2023-01-10 at 09 04 57

## Checklist before requesting a review

- [ ] I have performed a self-review of my code

[GRW-1998]: https://hedvig.atlassian.net/browse/GRW-1998?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 1]
```diff
@@ -125,6 +130,31 @@ export const OfferPresenter = (props: Props) => {
 
   const startDate = convertToDate(selectedOffer.startDate) ?? new Date()
 
+  const toggleComparisonTable = () => {
+    setIsComparisonTableOpen(!isComparisonTableOpen)
+  }
+
+  const getAllPerils = () => {
```
Comment (alebedev) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:137]: Do you even need non-unique perils?  If not, just inline this into getUniquePerilTitles

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 2]
```diff
@@ -125,6 +130,31 @@ export const OfferPresenter = (props: Props) => {
...
+    return priceIntent.offers.reduce(
```
Comment (alebedev) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:138]: Try `flatMap`, should be easier

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 3]
```diff
@@ -161,6 +191,49 @@ export const OfferPresenter = (props: Props) => {
...
+      {isComparisonTableOpen && (
+        <ComparisonTable.Root>
+          <ComparisonTable.Head>
+            <ComparisonTable.Header />
+            {priceIntent.offers.map((offer) => (
+              <ComparisonTable.Header
+                key={offer.id}
+                active={isSelectedOffer(offer, selectedOfferId)}
+              >
+                {offer.variant.displayName}Half
```
Comment (alebedev) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:212]: Is "Half" temporary?

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 4]
```diff
@@ -161,6 +191,49 @@ export const OfferPresenter = (props: Props) => {
           </Space>
         </Space>
       </form>
+      {priceIntent.offers.length > 1 && (
+        <TextButton onClick={toggleComparisonTable}>
+          <StyledCrossIcon
+            transform={isComparisonTableOpen ? 'rotate(0)' : 'rotate(-45)'}
+            size="0.875rem"
+          />
+          Compare coverage
```
Comment (alebedev) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:200]: Localization?

**Review by alebedev — APPROVED** [review]
Looks good, some questions on minor details

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 5]
```diff
@@ -20,9 +21,12 @@ import { useTracking } from '@/services/Tracking/useTracking'
 import { convertToDate } from '@/utils/date'
 import { useFormatter } from '@/utils/useFormatter'
 import { CancellationForm, CancellationOption } from './CancellationForm/CancellationForm'
+import * as ComparisonTable from './ComparisonTable/ComparisonTable'
 import { PriceMatchBubble } from './PriceMatchBubble/PriceMatchBubble'
 import { useHandleSubmitAddToCart } from './useHandleSubmitAddToCart'
 
+const removeDuplicates = <T,>(arr: T[]): T[] => Array.from(new Set(arr))
```
Comment (alebedev) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:28]: `<T,>`?

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 6]
```diff
@@ -161,6 +191,49 @@ export const OfferPresenter = (props: Props) => {
           </Space>
         </Space>
       </form>
+      {priceIntent.offers.length > 1 && (
+        <TextButton onClick={toggleComparisonTable}>
```
Comment (robinandeer) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:195]: I think we have a "Collapsible" component that would work in this context and that takes care of all the a11y concerns... but we can address that in a follow-up PR

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 7]
```diff
@@ -125,6 +130,31 @@ export const OfferPresenter = (props: Props) => {
...
+    return removeDuplicates(perilTitles)
+  }
+
+  const offerHasPeril = (offer: ProductOfferFragment, perilTitle: string) =>
+    offer.variant.perils.some((peril) => peril.title === perilTitle)
+
+  const isSelectedOffer = (offer: ProductOfferFragment, selectedOfferId: string) => {
+    if (offer.id === selectedOfferId) return true
+    return false
+  }
```
Comment (robinandeer) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:156]: These can be moved outside the render function, no?

**Review by robinandeer — COMMENTED** [review]
Good stuff -- left some improvements suggestions :D

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 8]
```diff
@@ -200,6 +273,22 @@ const Separator = styled.div(({ theme }) => ({
   alignSelf: 'stretch',
 }))
 
+const TextButton = styled.button(({ theme }) => ({
```
[the remark this answers is not in the document]
Comment (amandaflod) [reply apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:276]: Checked with him and changing it to ghost! :))

File: apps/store/src/components/TierSelector/TierSelector.stories.tsx [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import { ComponentMeta, ComponentStory } from '@storybook/react'
 import { useState } from 'react'
 import { CurrencyCode, ExternalInsuranceCancellationOption } from '@/services/apollo/generated'
+import { mockedPerils } from '../Perils/Perils.stories'
```
[the remark this answers is not in the document]
Comment (amandaflod) [reply apps/store/src/components/TierSelector/TierSelector.stories.tsx:4]: We have a mocks-folder, a utils-folder and a test-folder, a bit unsure where to put it 😅

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx (hunk 3 above)
Comment (amandaflod) [reply apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx]: Yes!

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx (hunk 2 above)
Comment (amandaflod) [reply apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:146]: Great suggestion! Have fixed both of these now :)

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx (hunk 5 above)
Comment (amandaflod) [reply apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:27]: I wanted to use a generic type and found this workaround for this case https://stackoverflow.com/questions/32308370/what-is-the-syntax-for-typescript-arrow-functions-with-generics
I am open to other suggestions! :))

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 10]
```diff
@@ -125,44 +137,102 @@ export const OfferPresenter = (props: Props) => {
...
-      </form>
+        </form>
+        {priceIntent.offers.length > 1 && (
+          <SpaceFlex direction="vertical" align="center">
+            <Button variant="ghost" size="small" onClick={toggleComparisonTable}>
+              <StyledCrossIcon
+                transform={isComparisonTableOpen ? 'rotate(0)' : 'rotate(-45)'}
+                size="0.875rem"
+              />
+              {t('COMPARE_COVERAGE_BUTTON')}{' '}
```
Comment (alebedev) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:196]: why space?

**Review by alebedev — APPROVED** [review]
Looks good

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 11]
```diff
@@ -32,13 +35,22 @@ type Props = {
...
+  if (offer.id === selectedOfferId) return true
+  return false
+}
+
 export const OfferPresenter = (props: Props) => {
   const { priceIntent, shopSession, scrollPastRef, onAddedToCart, onClickEdit } = props
   const { t } = useTranslation('purchase-form')
   const formatter = useFormatter()
   const [selectedTypeOfContract, setSelectedTypeOfContract] = useState(
     priceIntent.offers[0].variant.typeOfContract,
```
Comment (robinandeer) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:51]: This is a placeholder, right? That should be based on the selected offer

We can fix it later, but add a TODO comment so we don't forget

File: apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx [hunk 12]
```diff
@@ -125,44 +137,102 @@ export const OfferPresenter = (props: Props) => {
...
+        {isComparisonTableOpen && (
+          <ComparisonTable.Root>
+            <ComparisonTable.Head>
+              <ComparisonTable.Header />
+              {priceIntent.offers.map((offer) => (
+                <ComparisonTable.Header
+                  key={offer.id}
+                  active={isSelectedOffer(offer, selectedOfferId)}
+                >
+                  {offer.variant.displayName}Trafik
```
Comment (robinandeer) [inline apps/store/src/components/ProductPage/PurchaseForm/OfferPresenter.tsx:210]: Add a TODO comment here as well

**Review by robinandeer — APPROVED** [review]
Good stuff!

**Review by guilhermespopolin — APPROVED** [review]
Well done 👏🏻

**Pull request merged**